src/theory/arith/theory_arith.cpp

```diff
@@ -73,6 +80,10 @@ ProofRuleChecker* TheoryArith::getProofChecker()
 
 bool TheoryArith::needsEqualityEngine(EeSetupInfo& esi)
 {
+  if (d_eqSolver != nullptr)
+  {
+    return d_eqSolver->needsEqualityEngine(esi);
```

Comment:
It looks a bit weird, that `d_internal` isn't even considered when `d_eqSolver` is used. It is correct, though, as `d_eqSolver` always returns `true`.
I'd still suggest to either formulate it like
`if (d_eqSolver != nullptr && d_eqSolver->needsEqualityEngine(esi)) return true` or just say `if (d_eqSolver != nullptr) return true;` to make this piece of code easier to understand.

Reply:
The purpose of this method is to initialize esi; I'll comment on why the control flow looks like this.